File: banking_tools/bank.py
"""Bank management system."""

import json
import os
from typing import Optional, List, Dict
from .account import Account


class Bank:
    """Main bank class managing multiple accounts."""
    
    def __init__(self, data_file: str = "bank_data.json"):
        """
        Initialize the bank.
        
        Args:
            data_file: Path to the data persistence file
        """
        self.data_file = data_file
        self.accounts: Dict[str, Account] = {}
        self.load_data()
    
    def load_data(self) -> None:
        """Load account data from file."""
        if os.path.exists(self.data_file):
            try:
                with open(self.data_file, 'r') as f:
                    data = json.load(f)
                    for acc_data in data:
                        account = Account.from_dict(acc_data)
                        self.accounts[account.account_number] = account
            except (json.JSONDecodeError, KeyError):
                # If file is corrupted, start fresh
                self.accounts = {}
    
    def save_data(self) -> None:
        """Save account data to file."""
        data = [account.to_dict() for account in self.accounts.values()]
        with open(self.data_file, 'w') as f:
            json.dump(data, f, indent=2)
    
    def create_account(self, account_number: str, account_holder: str,
                      password: str, initial_balance: float = 0.0) -> bool:
        """
        Create a new bank account.
        
        Args:
            account_number: Unique account identifier
            account_holder: Name of the account holder
            password: Account password
            initial_balance: Starting balance
            
        Returns:
            True if account created successfully, False if account already exists
        """
        if account_number in self.accounts:
            return False
        
        if initial_balance < 0:
            return False
        
        account = Account(account_number, account_holder, password, initial_balance)
        self.accounts[account_number] = account
        self.save_data()
        return True
    
    def authenticate(self, account_number: str, password: str) -> Optional[Account]:
        """
        Authenticate and return account.
        
        Args:
            account_number: Account identifier
            password: Account password
            
        Returns:
            Account object if authentication successful, None otherwise
        """
        account = self.accounts.get(account_number)
        if account and account.verify_password(password):
            return account
        return None
    
    def get_account(self, account_number: str) -> Optional[Account]:
        """Get account by account number (without authentication)."""
        return self.accounts.get(account_number)
    
    def transfer(self, from_account: str, to_account: str, amount: float,
                password: str) -> tuple[bool, str]:
        """
        Transfer money between accounts.
        
        Args:
            from_account: Source account number
            to_account: Destination account number
            amount: Amount to transfer
            password: Source account password
            
        Returns:
            Tuple of (success: bool, message: str)
        """
        # Authenticate source account
        source = self.authenticate(from_account, password)
        if not source:
            return False, "Authentication failed"
        
        # Check if destination account exists
        destination = self.get_account(to_account)
        if not destination:
            return False, "Destination account not found"
        
        # Check if source has sufficient balance
        if amount <= 0:
            return False, "Invalid amount"
        
        if source.balance < amount:
            return False, "Insufficient balance"
        
        # Perform transfer
        source.withdraw(amount, f"Transfer to {to_account}")
        destination.deposit(amount, f"Transfer from {from_account}")
        self.save_data()
        
        return True, "Transfer successful"
    
    def list_accounts(self) -> List[Dict]:
        """Get list of all accounts (basic info only)."""
        return [
            {
                "account_number": acc.account_number,
                "account_holder": acc.account_holder,
                "balance": acc.balance
            }
            for acc in self.accounts.values()
        ]

File: src/main/java/com/banking/Bank.java
package com.banking;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;

import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.lang.reflect.Type;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Main bank class managing multiple accounts.
 */
public class Bank {
    private String dataFile;
    private Map<String, Account> accounts;
    private Gson gson;

    /**
     * Initialize the bank.
     * 
     * @param dataFile Path to the data persistence file
     */
    public Bank(String dataFile) {
        this.dataFile = dataFile;
        this.accounts = new HashMap<>();
        this.gson = new GsonBuilder().setPrettyPrinting().create();
        loadData();
    }

    /**
     * Load account data from file.
     */
    private void loadData() {
        if (Files.exists(Paths.get(dataFile))) {
            try (FileReader reader = new FileReader(dataFile)) {
                Type listType = new TypeToken<ArrayList<Account>>(){}.getType();
                List<Account> accountList = gson.fromJson(reader, listType);
                if (accountList != null) {
                    for (Account account : accountList) {
                        accounts.put(account.getAccountNumber(), account);
                    }
                }
            } catch (IOException e) {
                // If file is corrupted, start fresh
                accounts = new HashMap<>();
            }
        }
    }

    /**
     * Save account data to file.
     */
    public void saveData() {
        try (FileWriter writer = new FileWriter(dataFile)) {
            List<Account> accountList = new ArrayList<>(accounts.values());
            gson.toJson(accountList, writer);
        } catch (IOException e) {
            System.err.println("Error saving data: " + e.getMessage());
        }
    }

    /**
     * Create a new bank account.
     * 
     * @param accountNumber Unique account identifier
     * @param accountHolder Name of the account holder
     * @param password Account password
     * @param initialBalance Starting balance
     * @return true if account created successfully, false otherwise
     */
    public boolean createAccount(String accountNumber, String accountHolder, String password, double initialBalance) {
        if (accountNumber == null || accountNumber.trim().isEmpty()) {
            return false;
        }

        if (accountHolder == null || accountHolder.trim().isEmpty()) {
            return false;
        }

        if (accounts.containsKey(accountNumber)) {
            return false;
        }

        if (initialBalance < 0) {
            return false;
        }

        try {
            Account account = new Account(accountNumber, accountHolder, password, initialBalance);
            accounts.put(accountNumber, account);
            saveData();
            return true;
        } catch (IllegalArgumentException e) {
            // Password validation failed
            return false;
        }
    }

    /**
     * Authenticate and return account.
     * 
     * @param accountNumber Account identifier
     * @param password Account password
     * @return Account object if authentication successful, null otherwise
     */
    public Account authenticate(String accountNumber, String password) {
        Account account = accounts.get(accountNumber);
        if (account != null && account.verifyPassword(password)) {
            return account;
        }
        return null;
    }

    /**
     * Get account by account number (without authentication).
     * 
     * @param accountNumber Account identifier
     * @return Account object or null
     */
    public Account getAccount(String accountNumber) {
        return accounts.get(accountNumber);
    }

    /**
     * Transfer money between accounts.
     * 
     * @param fromAccount Source account number
     * @param toAccount Destination account number
     * @param amount Amount to transfer
     * @param password Source account password
     * @return TransferResult with success status and message
     */
    public TransferResult transfer(String fromAccount, String toAccount, double amount, String password) {
        // Prevent self-transfer
        if (fromAccount.equals(toAccount)) {
            return new TransferResult(false, "Cannot transfer to the same account");
        }

        // Authenticate source account
        Account source = authenticate(fromAccount, password);
        if (source == null) {
            return new TransferResult(false, "Authentication failed");
        }

        // Check if destination account exists
        Account destination = getAccount(toAccount);
        if (destination == null) {
            return new TransferResult(false, "Destination account not found");
        }

        // Check if source has sufficient balance
        if (amount <= 0) {
            return new TransferResult(false, "Invalid amount");
        }

        if (source.getBalance() < amount) {
            return new TransferResult(false, "Insufficient balance");
        }

        // Perform transfer
        source.withdraw(amount, "Transfer to " + toAccount);
        destination.deposit(amount, "Transfer from " + fromAccount);
        saveData();

        return new TransferResult(true, "Transfer successful");
    }

    /**
     * Get list of all accounts (basic info only).
     * 
     * @return List of account information maps
     */
    public List<Map<String, Object>> listAccounts() {
        List<Map<String, Object>> accountList = new ArrayList<>();
        for (Account account : accounts.values()) {
            Map<String, Object> info = new HashMap<>();
            info.put("account_number", account.getAccountNumber());
            info.put("account_holder", account.getAccountHolder());
            info.put("balance", account.getBalance());
            accountList.add(info);
        }
        return accountList;
    }

    /**
     * Result class for transfer operations.
     */
    public static class TransferResult {
        private boolean success;
        private String message;

        public TransferResult(boolean success, String message) {
            this.success = success;
            this.message = message;
        }

        public boolean isSuccess() {
            return success;
        }

        public String getMessage() {
            return message;
        }
    }
}
